lib/infrastructure/auth/firebase_auth_repository.dart

```diff
@@ -0,0 +1,202 @@
+
+import '../../domain/auth/auth_failures.dart';
+import '../../domain/auth/auth_success.dart';
+import '../../domain/auth/i_auth_repository.dart';
+import '../../domain/user/i_user_repository.dart';
+import '../../domain/user/user.dart';
+import 'firebase_auth_mapper.dart';
+
+@LazySingleton(as: IAuthRepository)
+class FirebaseAuthRepository implements IAuthRepository, Disposable {
+  final fb_auth.FirebaseAuth firebaseAuth;
+  final _userSubject = BehaviorSubject<User>.seeded(User.anonymous);
+  late final StreamSubscription _userStreamSubscription;
+
+  FirebaseAuthRepository({required this.firebaseAuth}) {
+    _userStreamSubscription = firebaseAuth
+        .authStateChanges()
+        .map(_firebaseUserToUser)
+        .listen(_userSubject.sink.add);
+  }
+
+  @override
+  Future<Option<User>> getSignedInUser() async {
+    return optionOf(firebaseAuth.currentUser?.toDomain());
+  }
+
+  @override
+  Future<void> signOut() => firebaseAuth.signOut();
+
+  @override
+  Stream<Either<AuthFailure, AuthSuccess>> verifyPhone(
+      {required String phoneNumber}) {
+    final result = StreamController<Either<AuthFailure, AuthSuccess>>();
+
+    var credential = const Credential();
+
+    firebaseAuth.verifyPhoneNumber(
+      phoneNumber: "+$phoneNumber",
+      codeSent: (String verificationId, int? forceResendingToken) {
+        credential = credential.copyWith(
+          verificationId: verificationId,
+          forceResendToken: forceResendingToken,
+        );
+
+        result.add(right(AuthSuccess.codeSent(credential: credential)));
+      },
+      verificationFailed: (fb_auth.FirebaseAuthException error) {
+        result.add(left(error.toFailure()));
+        result.close();
+      },
+      codeAutoRetrievalTimeout: (String verificationId) {
+        result.add(right(AuthSuccess.codeRetrievalTimedOut(
+            credential: credential.copyWith(verificationId: verificationId))));
+        result.close();
+      },
+      verificationCompleted: (fb_auth.PhoneAuthCredential phoneAuthCredential) {
+        credential = credential.copyWith(
+          verificationId: phoneAuthCredential.verificationId,
+          smsCode: phoneAuthCredential.smsCode,
+        );
+
+        result.add(
+            right(AuthSuccess.verificationCompleted(credential: credential)));
+        result.close();
+      },
+    );
+
+    return result.stream.distinct();
+  }
+
+  @override
+  Future<Either<AuthFailure, bool>> signInWithPhone(
+      {required Credential authCredentials}) async {
+    try {
+      final String verificationId = authCredentials.verificationId!;
+      final String smsCode = authCredentials.smsCode!;
+      // Create a PhoneAuthCredential with the code
+      final credential = fb_auth.PhoneAuthProvider.credential(
+          verificationId: verificationId, smsCode: smsCode);
+
+      // Sign the user in (or link) with the credential
+      await firebaseAuth.signInWithCredential(credential);
+      //TODO Check if is new user
```

Comment:
Do we need a new commit for this, or can we do it in this PR?

Reply:
Let me do that in this P.R